lib/jxl/modular/transform/enc_palette.cc

```diff
@@ -95,9 +96,69 @@ static int QuantizeColorToImplicitPaletteIndex(
 
 }  // namespace palette_internal
 
-Status FwdPalette(Image &input, uint32_t begin_c, uint32_t end_c,
-                  uint32_t &nb_colors, bool ordered, bool lossy,
-                  Predictor &predictor, const weighted::Header &wp_header) {
+int RoundInt(int value, int div) {  // symmetric rounding around 0
+  if (value < 0) return -RoundInt(-value, div);
+  return (value + div / 2) / div;
+}
+
+struct PaletteIterationData {
+  static constexpr int kMaxDeltas = 128;
+  bool final_run = false;
+  std::vector<pixel_type> deltas[3];
+  std::vector<double> delta_distances;
+  std::vector<pixel_type> frequent_deltas[3];
+
+  // Populates `frequent_deltas` with items from `deltas` based on frequencies
+  // and color distances.
+  void FindFrequentColorDeltas(int num_pixels) {
+    using pixel_type_3d = std::array<pixel_type, 3>;
+    std::map<pixel_type_3d, double> delta_frequency_map;
+    for (size_t i = 0; i < deltas[0].size();
+         ++i) {  // store frequency weighted by delta distance from quantized
+                 // value
+      pixel_type_3d delta = {
+          RoundInt(deltas[0][i], 3), RoundInt(deltas[1][i], 3),
+          RoundInt(deltas[2][i], 3)};  // a basic form of clustering
+      if (deltas[0][i] == 0 && deltas[1][i] == 0 && deltas[2][i] == 0) continue;
```

Comment:
Just checking, but is this (rather than `delta[0] == 0 && delta[1] == 0 && delta[2] == 0`) the correct check?

Reply:
It should be the same. Done.